File: App/Resources/AlgorithmDetails/iterativetopdownmergesort/iterativetopdownmergesort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void merge(int arr[], int low, int mid, int high) {
  int leftSize = mid - low;
  int rightSize = high - mid;
  int *left = malloc(leftSize * sizeof(int));
  int *right = malloc(rightSize * sizeof(int));
  for (int x = 0; x < leftSize; x++) {
    left[x] = arr[low + x];
  }
  for (int x = 0; x < rightSize; x++) {
    right[x] = arr[mid + x];
  }

  int i = 0, j = 0, k = low;
  while (i < leftSize && j < rightSize) {
    if (left[i] <= right[j]) {
      arr[k++] = left[i++];
    } else {
      arr[k++] = right[j++];
    }
  }
  while (i < leftSize) {
    arr[k++] = left[i++];
  }
  while (j < rightSize) {
    arr[k++] = right[j++];
  }

  free(left);
  free(right);
}

void sort(int arr[], int n) {
  int subarrayCount = 1;
  while (subarrayCount < n) {
    subarrayCount *= 2;
  }

  while (subarrayCount > 1) {
    for (int i = 0; i < subarrayCount; i += 2) {
      int low = n * i / subarrayCount;
      int mid = n * (i + 1) / subarrayCount;
      int high = n * (i + 2) / subarrayCount;
      merge(arr, low, mid, high);
    }
    subarrayCount /= 2;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: Modules/BuiltInAlgorithms/Sources/Merge/IterativeTopDownMergeSort.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `IterativeTopDownMergeSort` — an ordinary top-down merge sort (each merge
/// splits its range at the true midpoint, recursively), but computed level-by-level with plain
/// arithmetic instead of recursion. Starting from `subarrayCount` = the smallest power of two at
/// least `n`, each pass merges adjacent proportional slices (`n*i/subarrayCount` boundaries) and
/// halves `subarrayCount`; because the boundaries are proportional rather than fixed-width, this
/// naturally produces the same balanced split a recursive top-down merge sort would even when `n`
/// isn't a power of two — no separate fixup pass needed for the remainder, unlike
/// `BottomUpMergeSort`'s fixed-doubling-width approach.
///
/// ArrayV's own `runSortLarge` variant (rational-number arithmetic standing in for the
/// multiplication/division above, to dodge overflow at very large `n`) only triggers past
/// `length >= 1 << 15` — this app caps every algorithm's `sizeRange` at 256, so that branch is
/// unreachable here and isn't ported.
///
/// Stable by construction, not by fuzzing: the merge step only ever takes from the right run
/// (`high`) when the left run's current element compares strictly greater (`by: (>)`), so on a tie
/// it always takes from the left run first — the standard stable-merge tie-break.
public struct IterativeTopDownMergeSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "iterativetopdownmergesort")
  public let metadata = AlgorithmMetadata(
    displayName: "Iterative Top-Down Merge Sort",
    category: .merge,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 2390, coefficients: [188980, 92.2662, 0.00294775],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .powerLog, coefficients: [7.54425, 1.03832], rSquared: 0.999756),
    implementationComplexity: 15,
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(n)",
    iconName: "square.stack.3d.down.forward"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n >= 2 else { return }

    let tempHandle = engine.createAuxArray(length: n)
    var scratch = engine.values

    func merge(_ start: Int, _ mid: Int, _ end: Int) {
      var low = start
      var high = mid
      var nxt = start

      func take(from source: Int) {
        scratch[nxt] = engine.values[source]
        engine.writeAux(tempHandle, at: nxt, value: engine.values[source])
      }

      while low < mid && high < end {
        if engine.compare(low, high, by: (>)) {
          take(from: high)
          high += 1
        } else {
          take(from: low)
          low += 1
        }
        nxt += 1
      }
      if low >= mid {
        while high < end {
          take(from: high)
          high += 1
          nxt += 1
        }
      } else {
        while low < mid {
          take(from: low)
          low += 1
          nxt += 1
        }
      }

      for i in start..<end {
        engine.setValue(i, scratch[i])
      }
    }

    func ceilPowerOfTwo(_ x: Int) -> Int {
      var x = x - 1
      var i = 16
      while i > 0 {
        x |= x >> i
        i >>= 1
      }
      return x + 1
    }

    var subarrayCount = ceilPowerOfTwo(n)
    while subarrayCount > 1 {
      var i = 0
      while i < subarrayCount {
        merge(n * i / subarrayCount, n * (i + 1) / subarrayCount, n * (i + 2) / subarrayCount)
        i += 2
      }
      subarrayCount >>= 1
    }

    engine.deleteAuxArray(tempHandle)
  }
}
